File: bindings/lua/tests/smoke.lua
#!/usr/bin/env luajit
-- SPDX-License-Identifier: 0BSD

package.path = "./?.lua;" .. package.path

local rrc = require("rrc")

if rrc.version() ~= rrc.API_VERSION then
	io.stderr:write("version mismatch\n")
	os.exit(1)
end

local node = rrc.node_create("")
rrc.node_start(node)
rrc.node_stop(node)
rrc.node_destroy(node)

local id = rrc.identity_generate()
if #rrc.identity_hash(id) ~= rrc.HASH_LEN then
	io.stderr:write("bad identity hash\n")
	os.exit(1)
end
rrc.identity_destroy(id)

local sender = string.char(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
local env = rrc.envelope_create(rrc.RRC_TYPE_MSG, sender)
rrc.envelope_set_room(env, "lobby")
rrc.envelope_set_body_text(env, "hello")
local data = rrc.envelope_marshal(env)
rrc.envelope_destroy(env)

local got = rrc.envelope_unmarshal(data)
if rrc.envelope_body_text(got) ~= "hello" then
	io.stderr:write("body mismatch\n")
	os.exit(1)
end
rrc.envelope_destroy(got)

if rrc.normalize_room("  #Lobby ") ~= "#lobby" then
	io.stderr:write("normalize room failed\n")
	os.exit(1)
end

print("lua-smoke ok")

File: bindings/rust/tests/smoke.rs
// SPDX-License-Identifier: 0BSD

use rrc::{version, Envelope, Identity, Node, API_VERSION, HASH_LEN, RRC_TYPE_MSG};

#[test]
fn version_matches_api() {
    assert_eq!(version(), API_VERSION);
}

#[test]
fn envelope_roundtrip() {
    let sender: Vec<u8> = (0..HASH_LEN).map(|i| i as u8).collect();
    let env = Envelope::create(RRC_TYPE_MSG, &sender).expect("create");
    env.set_room("lobby").expect("room");
    env.set_body_text("hello").expect("body");
    let data = env.marshal().expect("marshal");

    let got = Envelope::unmarshal(&data).expect("unmarshal");
    assert_eq!(got.body_text().expect("text"), "hello");
}

#[test]
fn node_lifecycle() {
    let node = Node::create("").expect("node");
    node.start().expect("start");
    node.stop().expect("stop");
}

#[test]
fn identity_hash() {
    let id = Identity::generate().expect("generate");
    assert_eq!(id.hash_bytes().expect("hash").len(), HASH_LEN);
}

#[test]
fn normalize_sanitize() {
    assert_eq!(rrc::normalize_room("  #Lobby ").expect("room"), "#lobby");
    assert_eq!(rrc::sanitize_nick(" alice ").expect("nick"), "alice");
}
